### PR TITLE
OTT-1043: fix fv in logger when floor rule not found

### PR DESCRIPTION
**Pull request by Pubmatic-Supriya-Patil**
…is matched and default value of model is selected as bidfloor.

# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: floors/rule.go [hunk 1]
```diff
@@ -116,9 +116,13 @@ func updateImpExtWithFloorDetails(imp *openrtb_ext.ImpWrapper, matchedRule strin
...
 	}
+	floorRuleValue := 0.0
+	if matchedRule != "" {
+		floorRuleValue = floorRuleVal
+	}
 	extImpPrebid.Floors = &openrtb_ext.ExtImpPrebidFloors{
 		FloorRule:      matchedRule,
-		FloorRuleValue: math.Floor(floorRuleVal*10000) / 10000,
+		FloorRuleValue: math.Floor(floorRuleValue*10000) / 10000,
 		FloorValue:     floorVal,
```
Comment (pm-nilesh-chate) [inline floors/rule.go:126]: should add test case to assert this behaviour for future changes. This applies for the above file changes as well.

**Review by pm-nilesh-chate — CHANGES_REQUESTED** [review]
nitpick: Title should be short and descriptive.
Ex. OTT-1043: fix fv in logger when floor rule not found

File: exchange/floors.go [hunk 2]
```diff
@@ -205,15 +205,7 @@ func updateBidExt(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids map[op
...
 				continue
 			}
-
-			reqImpCur := reqImp.BidFloorCur
-			if reqImpCur == "" {
-				reqImpCur = "USD"
-				if bidRequestWrapper.Cur != nil {
-					reqImpCur = bidRequestWrapper.Cur[0]
-				}
-			}
```
Comment (pm-nilesh-chate) [inline exchange/floors.go:215]: update tests for this change. Ex. `TestUpdateBidExt`

**Review by pm-nilesh-chate — APPROVED** [review]
Approved. do not merge until master drop is finalised

File: floors/rule.go (hunk 1 above)
Comment (Pubmatic-Supriya-Patil) [reply floors/rule.go:126]: Added UT's.

File: exchange/floors.go [hunk 3]
```diff
@@ -64,18 +64,23 @@ func updateBidExtWithFloors(reqImp *openrtb_ext.ImpWrapper, bid *entities.PbsOrt
...
+	var bidExtFloors openrtb_ext.ExtBidFloors
 
 	prebidExt := impExt.GetPrebid()
-	if prebidExt == nil || prebidExt.Floors == nil {
+	if prebidExt != nil && prebidExt.Floors != nil {
+		bidExtFloors.FloorRule = prebidExt.Floors.FloorRule
+		bidExtFloors.FloorRuleValue = prebidExt.Floors.FloorRuleValue
+		bidExtFloors.FloorValue = prebidExt.Floors.FloorValue
+		bidExtFloors.FloorCurrency = floorCurrency
+		bid.BidFloors = &bidExtFloors
```
Comment (pm-nilesh-chate) [inline exchange/floors.go:75]: bid.BidFloors = &openrtb_ext.ExtBidFloors{
    FloorRule: prebidExt.Floors.FloorRule,
    FloorRuleValue: prebidExt.Floors.FloorRuleValue,
    FloorValue: prebidExt.Floors.FloorValue,
    FloorCurrency: floorCurrency,
}

File: exchange/floors.go [hunk 4]
```diff
@@ -64,18 +64,23 @@ func updateBidExtWithFloors(reqImp *openrtb_ext.ImpWrapper, bid *entities.PbsOrt
...
 		return
+	} else {
+		if reqImp.Imp.BidFloor != 0 {
+			bidExtFloors.FloorValue = reqImp.Imp.BidFloor
+			bidExtFloors.FloorCurrency = floorCurrency
+			bid.BidFloors = &bidExtFloors
```
Comment (pm-nilesh-chate) [inline exchange/floors.go:81]:
```
bid.BidFloors = &openrtb_ext.ExtBidFloors{
    FloorValue: prebidExt.Floors.FloorValue,
    FloorCurrency: floorCurrency,
}
```

File: exchange/floors.go [hunk 5]
```diff
@@ -206,3 +211,23 @@ func enforceFloors(r *AuctionRequest, seatBids map[openrtb_ext.BidderName]*entit
...
+	}
+
+	for _, seatBid := range seatBids {
+
+		for _, bid := range seatBid.Bids {
+			reqImp, ok := impMap[bid.Bid.ImpID]
+			if !ok {
+				continue
+			}
+			updateBidExtWithFloors(reqImp, bid, reqImp.BidFloorCur)
```
Comment (pm-nilesh-chate) [inline exchange/floors.go:230]: currency UT cases should be part of `updateBidExtWithFloors`. `updateBidExt` is not responsible for changes to currency. Move the test cases accordingly

File: exchange/floors.go [hunk 6]
```diff
@@ -206,3 +211,23 @@ func enforceFloors(r *AuctionRequest, seatBids map[openrtb_ext.BidderName]*entit
...
+	for _, v := range bidRequestWrapper.GetImp() {
+		impMap[v.ID] = v
+	}
+
+	for _, seatBid := range seatBids {
+
+		for _, bid := range seatBid.Bids {
+			reqImp, ok := impMap[bid.Bid.ImpID]
+			if !ok {
+				continue
```
Comment (pm-nilesh-chate) [inline exchange/floors.go:228]: please cover impID to bid.impId UT cases

File: exchange/floors.go (hunk 5 above)
Comment (pm-nilesh-chate) [reply exchange/floors.go:230]: Not need to pass `reqImp.BidFloorCur` separately

**Review by pm-nilesh-chate — CHANGES_REQUESTED** [review]
requested minor refactor

File: exchange/floors.go [hunk 7]
```diff
@@ -58,24 +58,30 @@ func floorsEnabled(account config.Account, bidRequestWrapper *openrtb_ext.Reques
...
-	if prebidExt == nil || prebidExt.Floors == nil {
+	if prebidExt != nil && prebidExt.Floors != nil {
+		bid.BidFloors = &openrtb_ext.ExtBidFloors{
+			FloorRule:      prebidExt.Floors.FloorRule,
+			FloorRuleValue: prebidExt.Floors.FloorRuleValue,
+			FloorValue:     prebidExt.Floors.FloorValue,
+			FloorCurrency:  reqImp.BidFloorCur,
+		}
 		return
+	} else {
```
Comment (pm-viral-vala) [inline exchange/floors.go:77]: There is no need to have else block as we are returning from if block as is.

**Review by pm-viral-vala — CHANGES_REQUESTED** [review]

**Review by pm-viral-vala — APPROVED** [review]

**Pull request merged**